{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c3058d47-b12b-4b97-8cee-00627719ba43",
   "metadata": {},
   "source": [
    "# Fold first, ask later: structure-informed function annotation of *Pseudomonas* phage proteins\n",
    "## Pipeline stage 2: annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfc12ff8-a7e5-42d8-80f1-2758c61d0b6a",
   "metadata": {},
   "source": [
    "### Stage 2.X: annotation - investigating a future-proof annotation classification method\n",
    "\n",
    "**Goals:**\n",
    "* Identify a feasible, representative subset of protein annotations to train the classifier on.\n",
    "* Label using the current REGEX method.\n",
    "* Manually correct labels to desired label.\n",
    "\n",
    "**Requires:**\n",
    "* in folder Desktop/phage_annotation_input:\n",
    "    * FoldSeek search database header files af50m_2024mar_h (FoldSeek 'Alphafold/UniProt50-minimal' database v.4) \n",
    "\n",
    "**Generates:**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9244f087-5325-43c2-bf89-8706366a78d8",
   "metadata": {},
   "source": [
    "#### General settings, imports, variables and environments\n",
    "\n",
    "Conda environment: ffal_annotate\n",
    "\n",
    "Created with `conda create -n ffal_annotate`. Then installed `jupyter notebook`, `pandas` through conda & `seaborn`, `scipy`, `matplotlib-venn`, `UpSetPlot`, `biopython` through pip (commands executed in notebook Code_annotation_background.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d4cc488-01d8-41b9-abe2-9c719065beb2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!pip install seaborn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f252af7-042a-4f07-89a2-b645f5c76b80",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e391fa57-c120-4e32-90ba-92fb8fee42b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!pip install matplotlib-venn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36228803-3c84-4cce-b3f6-510a12ea409c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!pip install UpSetPlot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2720272-db5a-4953-8221-a9cbcaa9c2c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bd9f9f1a-3d0d-40b8-8909-31dc90a18a9c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# This file may be used to create an environment using:\n",
      "# $ conda create --name <env> --file <this file>\n",
      "# platform: win-64\n",
      "# created-by: conda 24.11.3\n",
      "@EXPLICIT\n",
      "https://conda.anaconda.org/conda-forge/noarch/ca-certificates-2025.6.15-h4c7d964_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/intel-openmp-2024.2.1-h57928b3_1083.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/python_abi-3.13-7_cp313.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/tzdata-2025b-h78e105d_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/ucrt-10.0.22621.0-h57928b3_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/winpty-0.4.3-4.tar.bz2\n",
      "https://conda.anaconda.org/conda-forge/win-64/libwinpthread-12.0.0.r4.gg4f2fc60ca-h57928b3_9.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/vc14_runtime-14.44.35208-h818238b_26.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/vc-14.3-h41ae7f8_26.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/vs2015_runtime-14.44.35208-h38c0c73_26.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/bzip2-1.0.8-h2466b09_7.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libexpat-2.7.0-he0c23c2_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libffi-3.4.6-h537db12_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libiconv-1.18-h135ad9c_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/liblzma-5.8.1-h2466b09_2.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libmpdec-4.0.0-h2466b09_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libsodium-1.0.20-hc70643c_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libsqlite-3.50.1-hf5d6505_7.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libzlib-1.3.1-h2466b09_2.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/openssl-3.5.0-ha4e3fda_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/tk-8.6.13-h2c6b04d_2.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/yaml-0.2.5-h8ffe710_2.tar.bz2\n",
      "https://conda.anaconda.org/conda-forge/win-64/krb5-1.21.3-hdf4eb48_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libxml2-2.13.8-h442d1da_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/python-3.13.5-h7de537c_102_cp313.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/asttokens-3.0.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/attrs-25.3.0-pyh71513ae_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/brotli-python-1.1.0-py313h5813708_3.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/cached_property-1.5.2-pyha770c72_1.tar.bz2\n",
      "https://conda.anaconda.org/conda-forge/noarch/certifi-2025.6.15-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/charset-normalizer-3.4.2-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/colorama-0.4.6-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/cpython-3.13.5-py313hd8ed1ab_102.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/debugpy-1.8.14-py313h5813708_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/decorator-5.2.1-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/defusedxml-0.7.1-pyhd8ed1ab_0.tar.bz2\n",
      "https://conda.anaconda.org/conda-forge/noarch/executing-2.2.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/hpack-4.1.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/hyperframe-6.1.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/idna-3.10-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/json5-0.12.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/jsonpointer-3.0.0-py313hfa70ccb_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyterlab_widgets-3.0.15-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libhwloc-2.11.2-default_ha69328c_1001.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/markupsafe-3.0.2-py313hb4c8b1a_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/nest-asyncio-1.6.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/packaging-25.0-pyh29332c3_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pandocfilters-1.5.0-pyhd8ed1ab_0.tar.bz2\n",
      "https://conda.anaconda.org/conda-forge/noarch/parso-0.8.4-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pickleshare-0.7.5-pyhd8ed1ab_1004.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pip-25.1.1-pyh145f28c_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pkgutil-resolve-name-1.3.10-pyhd8ed1ab_2.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/platformdirs-4.3.8-pyhe01879c_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/prometheus_client-0.22.1-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/psutil-7.0.0-py313ha7868ed_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pure_eval-0.2.3-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pycparser-2.22-pyh29332c3_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pygments-2.19.2-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/python-fastjsonschema-2.21.1-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/python-json-logger-2.0.7-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/python-tzdata-2025.2-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pytz-2025.2-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/pywin32-307-py313h5813708_3.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/pywinpty-2.0.15-py313h5813708_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/pyyaml-6.0.2-py313hb4c8b1a_2.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/rfc3986-validator-0.1.1-pyh9f0ad1d_0.tar.bz2\n",
      "https://conda.anaconda.org/conda-forge/win-64/rpds-py-0.25.1-py313ha8a9a3c_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/setuptools-80.9.0-pyhff2d567_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/six-1.17.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/sniffio-1.3.1-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/soupsieve-2.7-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/tomli-2.2.1-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/tornado-6.5.1-py313ha7868ed_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/traitlets-5.14.3-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/types-python-dateutil-2.9.0.20250516-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/typing_extensions-4.14.0-pyhe01879c_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/typing_utils-0.1.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/uri-template-1.3.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/wcwidth-0.2.13-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/webcolors-24.11.1-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/webencodings-0.5.1-pyhd8ed1ab_3.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/websocket-client-1.8.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/widgetsnbextension-4.0.14-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/win_inet_pton-1.1.0-pyh7428d3b_8.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/zeromq-4.3.5-ha9f60a1_7.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/zipp-3.23.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/async-lru-2.0.5-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/babel-2.17.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/bleach-6.2.0-pyh29332c3_4.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/cached-property-1.5.2-hd8ed1ab_1.tar.bz2\n",
      "https://conda.anaconda.org/conda-forge/win-64/cffi-1.17.1-py313ha7868ed_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/comm-0.2.2-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/exceptiongroup-1.3.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/h11-0.16.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/h2-4.2.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/importlib-metadata-8.7.0-pyhe01879c_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/importlib_resources-6.5.2-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/ipython_pygments_lexers-1.1.1-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jedi-0.19.2-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jinja2-3.1.6-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter_core-5.8.1-pyh5737063_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyterlab_pygments-0.3.0-pyhd8ed1ab_2.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/matplotlib-inline-0.1.7-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/mistune-3.1.3-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/overrides-7.7.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/prompt-toolkit-3.0.51-pyha770c72_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/pysocks-1.7.1-pyh09c184e_7.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/python-dateutil-2.9.0.post0-pyhff2d567_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/pyzmq-27.0.0-py313h2100fd5_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/referencing-0.36.2-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/rfc3339-validator-0.1.4-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/send2trash-1.8.3-pyh5737063_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/stack_data-0.6.3-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/tbb-2021.13.0-h62715c5_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/terminado-0.18.1-pyh5737063_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/tinycss2-1.4.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/typing-extensions-4.14.0-h32cad80_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/anyio-4.9.0-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/argon2-cffi-bindings-21.2.0-py313ha7868ed_5.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/arrow-1.3.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/beautifulsoup4-4.13.4-pyha770c72_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/bleach-with-css-6.2.0-h82add2a_4.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/fqdn-1.5.1-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/ipython-9.3.0-pyh6be1c34_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jsonschema-specifications-2025.4.1-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter_client-8.6.3-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter_server_terminals-0.5.3-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/mkl-2024.2.2-h66d3029_15.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/prompt_toolkit-3.0.51-hd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/zstandard-0.23.0-py313ha7868ed_2.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/argon2-cffi-25.1.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/httpcore-1.0.9-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/ipykernel-6.29.5-pyh4bbf305_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/ipywidgets-8.1.7-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/isoduration-20.11.0-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jsonschema-4.24.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libblas-3.9.0-32_h641d27c_mkl.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/urllib3-2.5.0-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/httpx-0.28.1-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jsonschema-with-format-nongpl-4.24.0-hd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter_console-6.6.3-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/libcblas-3.9.0-32_h5e41251_mkl.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/liblapack-3.9.0-32_h1aa476e_mkl.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/nbformat-5.10.4-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/requests-2.32.4-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter_events-0.12.0-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/nbclient-0.10.2-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/numpy-2.3.0-py313hefb8edb_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/nbconvert-core-7.16.6-pyh29332c3_0.conda\n",
      "https://conda.anaconda.org/conda-forge/win-64/pandas-2.3.0-py313hf91d08e_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter_server-2.16.0-pyhe01879c_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter-lsp-2.2.5-pyhe01879c_2.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyterlab_server-2.27.3-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/notebook-shim-0.2.4-pyhd8ed1ab_1.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyterlab-4.4.3-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/notebook-7.4.3-pyhd8ed1ab_0.conda\n",
      "https://conda.anaconda.org/conda-forge/noarch/jupyter-1.1.1-pyhd8ed1ab_1.conda\n"
     ]
    }
   ],
   "source": [
    "!conda list --explicit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c27449c9-c25c-430b-bf54-aad844e93b72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports - just kept from annotation notebook, probably don't use all of those here\n",
    "import csv\n",
    "import json\n",
    "import os\n",
    "import random\n",
    "import re\n",
    "import requests\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import Bio\n",
    "\n",
    "from Bio.PDB import *\n",
    "from json.decoder import JSONDecodeError\n",
    "from scipy.stats import mannwhitneyu\n",
    "from upsetplot import plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9e085e73-7ae1-4931-b298-27d32fc656fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear reference to different directories\n",
    "pipeline_annot_dir = os.getcwd()\n",
    "pipeline_search_dir = os.path.join(pipeline_annot_dir, os.pardir, \"1_dataset\")\n",
    "master_dir = os.path.abspath(os.path.join(pipeline_annot_dir, os.pardir, os.pardir))\n",
    "data_dir = os.path.join(os.path.join(os.environ['USERPROFILE']), 'Desktop', 'phage_annotation_input')\n",
    "struct_out_dir = os.path.join(os.path.join(os.environ[\"USERPROFILE\"]), \"Desktop\", \"phage_annotation_input\", \"structure_prediction\") \n",
    "comp_out_dir = os.path.join(os.path.join(os.environ[\"USERPROFILE\"]), \"Desktop\", \"phage_annotation_input\", \"structure_comparison\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2631119-5725-4e5b-85d8-d548e604076c",
   "metadata": {},
   "source": [
    "#### Goal 1 : Identify a feasible, representative subset of protein annotations to train the classifier on.\n",
    "\n",
    "Let's try and see how much unique annotations are in the reduced AlphaFold database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96966cce-cb0e-4b8f-bf9b-e0e4ece4f1ae",
   "metadata": {},
   "outputs": [],
   "source": [
    " # read in the protein annotation data - AF50m database\n",
    "af50mid_descr_db = {}\n",
    "with open(os.path.join(data_dir, \"af50m_2024mar_h\")) as fp:\n",
    "    for line in fp:\n",
    "        if len(line.strip()) != 0 :\n",
    "            af_id = line.split()[0].removeprefix(\"\\x00\")\n",
    "            descr = \" \".join(line.split()[1:])\n",
    "            af50mid_descr_db[af_id] = descr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8ab4371-b04f-4f5d-8582-c601230d309e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create an overview of the unique annotations in the AFdb\n",
    "unique_ann = list(set(af50mid_descr_db.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4b080cef-33ce-43b5-a294-b8e2c0077922",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1618820"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_ann) # 1.6M, which is a lot - let's reduce this "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf1a8fd6-005a-4696-a1f9-55f1baa70a95",
   "metadata": {},
   "source": [
    "If we look at all the unique entries, we see there's 1.6 million protein entries. It is not feasible to manually label those right now. Let's try and focus on the subset that we actually found during the annotation process:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "becbe156-cee0-4b0c-bfcb-9092c86f4c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reading in the map between UniProt ID and UniProt protein names\n",
    "    # since I did not use DictWriter properly, we have created too large fields for DictReader to reader, which I first have to accomodate here\n",
    "max_int = sys.maxsize\n",
    "while True:\n",
    "    # decrease the max_int value by factor 10 \n",
    "    # as long as the OverflowError occurs.\n",
    "    try:\n",
    "        csv.field_size_limit(max_int)\n",
    "        break\n",
    "    except OverflowError:\n",
    "        max_int = int(max_int/10)\n",
    "csv.field_size_limit(max_int)\n",
    "    # actually reading in the data\n",
    "with open(os.path.join(data_dir, \"dict_UniProtID_function.csv\"), newline=\"\") as csvfile:\n",
    "    reader = csv.DictReader(csvfile)\n",
    "    for row in reader:\n",
    "        dict_UniProtID_function = row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6f26427-7a49-4eb6-8c03-b7f8e3c330bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# switching the dictionary keys from AFdb identifiers to UniProt identifiers\n",
    "af50m_up_descr = {key.split(\"-\")[1]: value for key, value in af50mid_descr_db.items() if \"AF-\" in key}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4019ed2d-c0c2-4af1-90a5-c994922bb54b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "None in af50m_up_descr.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "777523a2-d4e0-4028-a435-c3b4c4f07369",
   "metadata": {},
   "outputs": [],
   "source": [
    "# taking the subset for which we fetched information from UniProt\n",
    "af50mid_subset = {key: value for key, value in af50m_up_descr.items() if key in dict_UniProtID_function.keys()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "68d97f1c-6afd-485c-a580-c1318fe0d60f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "None in af50mid_subset.values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "20d2db0a-6db9-41a2-92a9-8fbecb6a6eb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting the unique annotations out of it\n",
    "unique_ann_sub = list(set(af50mid_subset.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "545536ef-36a7-44a6-ba9c-644301ee4a23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22103"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unique_ann_sub)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aefdf1b-996a-4a40-b111-73fcc6a480c4",
   "metadata": {},
   "source": [
    "#### Goal 2 : Label protein annotations using the current REGEX method.\n",
    "\n",
    "Let's now label these protein annotations, using the existing REGEXes:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbac6ecf-e659-4c12-aadf-93139b556449",
   "metadata": {},
   "source": [
    "For the AlphaFold database, three rounds of manual inspection of annotations led to the following REGEXes for filtering out uninformative annotations, in the context of phage proteins:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "57b8d06a-f136-42f1-8f22-e4f54f767269",
   "metadata": {},
   "outputs": [],
   "source": [
    "# final filter lists\n",
    "hypothetical_af_search = [\"whole genome shotgun sequence\"] \n",
    "hypothetical_af_match = [\"unnamed product\", \n",
    "                         r\"genome assembly, chromosome: [\\w+]+\", r\"str. [\\w+]+[\\d+]+\", \n",
    "                         \"protein of unknwon function\", \n",
    "                         \"conserved domain protein, histidine-rich\", r\"lin[\\d+]+ protein\", r\"orf[\\w+]+( domain-containing)?( protein)?\", \n",
    "                         \"phage related-protein\", \"phage d3 protein\", \n",
    "                         r\"dna, contig(: [\\w+]+)?\", r\"uncharacterized protein conserved in bacteria(, prophage-related)?\", r\"gene [\\d+]+ protein\",\n",
    "                         r\"gifsy-[\\d+] prophage protein\", r\"(hypothetical )?genomic island protein\",\n",
    "                         r\"(putative )?(conserved )?(\\([a-z ]+\\) )?((uncharacterized)|(hypothet(h)?ical)|(unannotated))( conserved)?( protein)?(, (pro)?phage-related)?(, isoform [AB])?\",\n",
    "                         r\"(predicted|unannotated|uncharacterized|hypothetical|putative( |_)?)?(expressed)?(conserved)?((mobile element)|(integron protein cassette))?(( )?domain)?(constituent)?([\\d+]+)?( |_)protein(_-_conserved)?\", \n",
    "                         r\"((hypothetical|putative|probable|conserved|uncharacterized) )?((hypothetical|putative|probable|conserved|uncharacterized) )?(bacterio|pro)?(phage)((-| )(like|related|associated))?( (hypothetical|putative|probable|conserved|uncharacterized))?( protein)?(,? [g]?p[\\d+]+\\b)?(, family)?(, putative)?\",\n",
    "                         r\"hk97( family)?( phage)? protein\", r\"putative similar to (bacterio)?phage protein\",\n",
    "                         r\"(putative )?uncharacterized protein(?! duf[\\d+]+)( \\w+)?\", \n",
    "                         r\"(phage( |-))?(protein )?(putative )?[g]?p[\\d+]+\\b(-like)?( domain(-containing)?)?(( |-)family)?( protein)?\"\n",
    "                        ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9cf9f612-15b4-41d5-aeba-3c5c03364145",
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions to check whether description is in the list of hypothetical protein annotations for AFdb annotations\n",
    "def is_hypothetical_af(descr):\n",
    "    hypothetical_label = False\n",
    "    for filter_descr_s in hypothetical_af_search:\n",
    "        if re.search(filter_descr_s, descr.strip().lower()) != None:\n",
    "            hypothetical_label = True\n",
    "    for filter_descr_m in hypothetical_af_match:\n",
    "        if re.fullmatch(filter_descr_m, descr.strip().lower()) != None:\n",
    "            hypothetical_label = True\n",
    "    return hypothetical_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1ae1d7be-b55b-40d0-88e9-e73bada7f29c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set of filters to identify low informative annotations\n",
    "low_af_search = [r\"duf[0-9]+\", \"br0599\"] \n",
    "low_af_match = [r\"uncharacterized (mitochondrial )?protein ((xf_1581/xf_1686)|(atmg00810-like))\",\n",
    "                  r\"uncharacterized( membrane)? protein ([\\w+]+ )?\\((upf)[\\d+]+( family\\))\", r\"cson[\\d+]+ protein\", \n",
    "                  \"putative phage-related exported protein\",\n",
    "                  r\"((hypothetical|predicted|putative) )?(conserved )?((pro)?phage )?(cytosolic|membrane|secreted|exported|cytoplasmic)( associated)?( phage)? protein(, putative)?\"\n",
    "                 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d6381e90-2989-4952-bfb2-dcaa5822d12a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to test whether an annotation is to be considered low informative\n",
    "def is_low_af(descr):\n",
    "    low_label = False\n",
    "    for filter_descr_s in low_af_search:\n",
    "        if re.search(filter_descr_s, descr.strip().lower()) != None:\n",
    "            low_label = True\n",
    "    for filter_descr_m in low_af_match:\n",
    "        if re.fullmatch(filter_descr_m, descr.strip().lower()) != None:\n",
    "            low_label = True\n",
    "    return low_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6b3a3001-fee9-4d72-bbe1-8d0e719967e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open(\"AF50m_annotations_REGEXlabelled.tsv\", \"w\")\n",
    "file.write(f\"protein_annotation\\tregex_label\\n\")\n",
    "for annotation in random.sample(unique_ann_sub, 22103):\n",
    "    label = \"\"\n",
    "    if is_hypothetical_af(annotation) == True:\n",
    "        label = \"uninformative\"\n",
    "    elif is_low_af(annotation) == True:\n",
    "        label = \"low informative\"\n",
    "    else:\n",
    "        label = \"proper\"\n",
    "    file.write(f\"{annotation}\\t{label}\\n\")\n",
    "file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b434f54a-be47-4921-9594-11e27f2f1a33",
   "metadata": {},
   "source": [
    "#### Goal 3 : Manually correct labels to desired label.\n",
    "\n",
    "Let's now manually add a third column to this table labeling the proteins with the desired label. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fce68e79-9fe0-403a-9d9b-9f98fbb9e2a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
