{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b9a57430",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3a271556",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Loading annotation and label file\n",
    "data = pd.read_csv(r'.\\AF50m_subset_REGEX_man_labels_5k.txt', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "ab2ac384",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_annotation</th>\n",
       "      <th>regex_label</th>\n",
       "      <th>manual_label</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NADPH-dependent 7-cyano-7-deazaguanine reducta...</td>\n",
       "      <td>proper</td>\n",
       "      <td>proper</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hydrogen peroxide-inducible genes activator &gt; ...</td>\n",
       "      <td>proper</td>\n",
       "      <td>proper</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Scoulerine-9-O-methyltransferase 1</td>\n",
       "      <td>proper</td>\n",
       "      <td>proper</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PadR domain-containing protein</td>\n",
       "      <td>proper</td>\n",
       "      <td>proper</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>protein mono-ADP-ribosyltransferase PARP9 isof...</td>\n",
       "      <td>proper</td>\n",
       "      <td>proper</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  protein_annotation regex_label manual_label  \\\n",
       "0  NADPH-dependent 7-cyano-7-deazaguanine reducta...      proper       proper   \n",
       "1  Hydrogen peroxide-inducible genes activator > ...      proper       proper   \n",
       "2                 Scoulerine-9-O-methyltransferase 1      proper       proper   \n",
       "3                     PadR domain-containing protein      proper       proper   \n",
       "4  protein mono-ADP-ribosyltransferase PARP9 isof...      proper       proper   \n",
       "\n",
       "  note  \n",
       "0  NaN  \n",
       "1  NaN  \n",
       "2  NaN  \n",
       "3  NaN  \n",
       "4  NaN  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "8dbd3a6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5000\n"
     ]
    }
   ],
   "source": [
    "# How many annotations have been labeled manually?\n",
    "print(data[\"manual_label\"].notna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ce9d6843",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5000"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Getting the annotatons that have been labeled manually\n",
    "annotation = data.loc[data[\"manual_label\"].notna(), [\"protein_annotation\"]]\n",
    "\n",
    "# checking that the right amount of annotations were selected\n",
    "len(annotation.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2f6f3678",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NADPH-dependent 7-cyano-7-deazaguanine reducta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hydrogen peroxide-inducible genes activator &gt; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Scoulerine-9-O-methyltransferase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PadR domain-containing protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>protein mono-ADP-ribosyltransferase PARP9 isof...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  protein_annotation\n",
       "0  NADPH-dependent 7-cyano-7-deazaguanine reducta...\n",
       "1  Hydrogen peroxide-inducible genes activator > ...\n",
       "2                 Scoulerine-9-O-methyltransferase 1\n",
       "3                     PadR domain-containing protein\n",
       "4  protein mono-ADP-ribosyltransferase PARP9 isof..."
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspecting the annotations\n",
    "annotation.head()\n",
    "\n",
    "# things are rough my man\n",
    "# so annotations contain numbers, dashes, colons...\n",
    "# uppercase terms to but i do think that's less important. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f546f9f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#preprocessing goals\n",
    "# normalize- lowercase and remove extra whitespace\n",
    "# oo should also remove commas\n",
    "# however, want to keep things like GO:0001234 together\n",
    "# consider stopwords\n",
    "\n",
    "# for colons- if in a word keep if end of word remove...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "37a6c3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "# hi thoughts\n",
    "# don't want to make _ blanspaces to maintain code identity\n",
    "# do want to make / blankspaces\n",
    "\n",
    "def clean_text(text):\n",
    "    text = text.lower()  # Lowercase\n",
    "    text = text.replace(\"/\", \" \")  # Replace slashes with spaces\n",
    "    text = re.sub(r\"\\s+\", \" \", text)  # remove multiple spaces\n",
    "    text = re.sub(r\"[^a-z0-9:\\-\\. ]\", \"\", text)  # keep alphanum, colon, dash, period\n",
    "    return text.strip()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "55435c31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generating cleaned annotations\n",
    "cleaned_annotations = annotation[\"protein_annotation\"].apply(clean_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "9be3067e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for tokenisation, want to keep things like GO terms with : together\n",
    "\n",
    "token_pattern = r\"[A-Za-z0-9:\\-\\.']+\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "7886d754",
   "metadata": {},
   "outputs": [],
   "source": [
    "#trying to figure out what stop words to use\n",
    "# considering this is a more nische context (biological) words may differ\n",
    "\n",
    "from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS\n",
    "\n",
    "# gets built in english stopwords from sklearn\n",
    "stopwords = list(ENGLISH_STOP_WORDS)\n",
    "\n",
    "# so my problem is what is a stopword in this context?\n",
    "# things that may appear frequently like \"uncharacterized\" is a signal, so I don't want to remove\n",
    "\n",
    "#lets try without any stopwords first\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f693b438",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('protein', 2533), ('putative', 601), ('family', 537), ('domain-containing', 439), ('domain', 268), ('phage', 261), ('dna', 257), ('subunit', 242), ('uncharacterized', 174), ('isoform', 165), ('tail', 131), ('regulator', 128), ('1', 119), ('factor', 117), ('2', 107), ('prophage', 105), ('transcriptional', 103), ('reductase', 101), ('helicase', 99), ('and', 87), ('type', 85), ('a', 81), ('hydrolase', 76), ('polymerase', 75), ('synthase', 74), ('of', 71), ('atpase', 67), ('kelch', 66), ('rna', 66), ('cell', 60)]\n"
     ]
    }
   ],
   "source": [
    "# stopwords 2.0\n",
    "\n",
    "# want to see the most common words in the annotations\n",
    "from collections import Counter\n",
    "\n",
    "def tokenize(text):\n",
    "    return re.findall(token_pattern, text.lower())\n",
    "\n",
    "all_tokens = [tok for text in cleaned_annotations for tok in tokenize(text)]\n",
    "token_counts = Counter(all_tokens)\n",
    "print(token_counts.most_common(30))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5941e301",
   "metadata": {},
   "outputs": [],
   "source": [
    "# so from this we find that protein is super common, buuuuut also putative\n",
    "# want to keep putative, might try not removing any stopwords for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "663889e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#so let's try TF-IDF\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "\n",
    "vectorizer = TfidfVectorizer(\n",
    "    token_pattern=token_pattern, # already have it\n",
    "    lowercase=True,\n",
    "    stop_words=stopwords, # only english atm\n",
    "    ngram_range=(1,2), # might want to increase\n",
    "    min_df=2, # words should appear in at least 2 annotations\n",
    "    max_df=0.8 # wods that appear in more than 80% of annotations are ignored\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "5ea91e09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['nadph-dependent',\n",
       " '7-cyano-7-deazaguanine',\n",
       " 'reductase',\n",
       " '7-cyano-7-deazaguanine reductase',\n",
       " 'hydrogen',\n",
       " 'peroxide-inducible',\n",
       " 'genes',\n",
       " 'activator',\n",
       " 'oxyr',\n",
       " 'hydrogen peroxide-inducible',\n",
       " 'peroxide-inducible genes',\n",
       " 'genes activator',\n",
       " 'activator oxyr',\n",
       " '1',\n",
       " 'padr',\n",
       " 'domain-containing',\n",
       " 'protein',\n",
       " 'domain-containing protein',\n",
       " 'isoform',\n",
       " 'x4',\n",
       " 'isoform x4',\n",
       " 'chromosome',\n",
       " 'replication',\n",
       " 'chromosome replication',\n",
       " 'g',\n",
       " 'u',\n",
       " 'mismatch-specific',\n",
       " 'dna',\n",
       " 'glycosylase',\n",
       " 'modular',\n",
       " 'u mismatch-specific',\n",
       " 'dna glycosylase',\n",
       " 'modular protein',\n",
       " 'p22',\n",
       " 'coat',\n",
       " '-',\n",
       " '5',\n",
       " 'family',\n",
       " 'p22 coat',\n",
       " 'protein 5',\n",
       " 'family protein',\n",
       " 'unplaced',\n",
       " 'genomic',\n",
       " 'scaffold',\n",
       " 'genome',\n",
       " 'shotgun',\n",
       " 'sequence',\n",
       " 'unplaced genomic',\n",
       " 'genomic scaffold',\n",
       " 'genome shotgun',\n",
       " 'shotgun sequence',\n",
       " 'kaic',\n",
       " 'putative',\n",
       " 'circadian',\n",
       " 'clock',\n",
       " 'circadian clock',\n",
       " 'clock protein',\n",
       " 'uncharacterized',\n",
       " 'x5',\n",
       " 'uncharacterized protein',\n",
       " 'isoform x5',\n",
       " 'rnase',\n",
       " 'transport',\n",
       " 'transport protein',\n",
       " 'f',\n",
       " 'acyl-coa',\n",
       " 'binding',\n",
       " 'acyl-coa binding',\n",
       " 'binding protein',\n",
       " 'bacteriophage',\n",
       " 'lambda',\n",
       " 'head',\n",
       " 'decoration',\n",
       " 'd',\n",
       " 'bacteriophage lambda',\n",
       " 'head decoration',\n",
       " 'decoration protein',\n",
       " 'protein d',\n",
       " 'transporter',\n",
       " 'cytosine-5-methyltransferase',\n",
       " '3',\n",
       " 'dna cytosine-5-methyltransferase',\n",
       " 'isoform 3',\n",
       " 'kelch',\n",
       " 'kelch domain-containing',\n",
       " 'bifunctional',\n",
       " 'oligoribonuclease',\n",
       " 'phosphatase',\n",
       " 'duf4376',\n",
       " 'domain',\n",
       " 'containing',\n",
       " 'duf4376 domain',\n",
       " 'domain containing',\n",
       " 'containing protein',\n",
       " 'recombination',\n",
       " 'mgsa',\n",
       " 'recombination protein',\n",
       " 'gnat-family',\n",
       " 'acetyltransferase',\n",
       " 'tigr03103']"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfidf = vectorizer.fit(cleaned_annotations)\n",
    "len(tfidf.vocabulary_)\n",
    "list(tfidf.vocabulary_.keys())[:100]  # Display the first 100 feature names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "23f89aca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['-' '0' '0 sporulation' '1' '1 isoform' '1 like' '1 member' '1 protein'\n",
      " '1 rna' '1 subunit' '1-like' '1-like isoform' '1-like protein'\n",
      " '1-phosphate' '1-phosphate phosphatase' '10' '11' '12' '12-dioxygenase'\n",
      " '13' '14' '15' '16' '17' '19' '1b' '2' '2 chloroplastic' '2 conserved'\n",
      " '2 domain' '2 isoform' '2 mitochondrial' '2 protein' '2-1'\n",
      " '2-1 chloroplastic' '2-dehydrogenase' '2-deoxyribosyltransferase'\n",
      " '2-like' '2-like 1' '2-like isoform' '2-oxoglutarate'\n",
      " '2-oxoglutarate 5-dioxygenase' '2-oxoglutarate iron-dependent' '20'\n",
      " '20 chloroplastic-like' '20s' '20s proteasome' '23' '23s' '23s rrna' '25'\n",
      " '26' '26s' '26s protease' '26s proteasome' '29' '29 isoform' '2a'\n",
      " '2og-feii' '2og-feii oxygenase' '3' '3 alpha' '3 isoform' '3-5'\n",
      " '3-5 exonuclease' '3-5 rna' '3-dehydrogenase' '3-like' '3-like protein'\n",
      " '3-oxoacyl-acyl-carrier-protein'\n",
      " '3-oxoacyl-acyl-carrier-protein reductase' '3-phosphatase'\n",
      " '3-pyrophosphohydrolase' '30' '31' '32' '33' '35' '3a-like' '4'\n",
      " '4 chloroplastic' '4-dehydrogenase' '4-epimerase' '4-hydroxylase'\n",
      " '4-like' '4-like protein' '40' '40 kda' '41' '42' '48' '48 homolog' '5'\n",
      " '5 isoform' '5 nucleotidase' '5-3' '5-3 exonuclease' '5-dehydrogenase'\n",
      " '5-deoxynucleotidase' '5-dioxygenase']\n"
     ]
    }
   ],
   "source": [
    "# inspecting some of the TF-IDF values\n",
    "feature_names = vectorizer.get_feature_names_out()\n",
    "print(feature_names[:100])  # Display the first 100 feature names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b865c24c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
